src/arraymancer/tensor/syntactic_sugar.nim

```diff
@@ -29,3 +29,51 @@ template at*[T](t: Tensor[T], args: varargs[untyped]): untyped =
   ## Usage:
   ##   See the ``[]`` macro
   t[args].squeeze
+
+template at_mut*[T](t: var Tensor[T], args: varargs[untyped]): untyped =
+  ## Slice a Tensor, collapse singleton dimension, returning a mutable slice of the input
+  ##
+  ## This can be useful, for example, when assigning a value into a chain
+  ## of slice operations which are usually considered immutable even if
+  ## the original tensor is mutable. For example, this lets you do:
+  ##
+  ## .. code:: nim
+  ##   var x = arange(12).reshape([4, 3])
+  ##   let condition = [[true, false, true], [true, false, true]].toTensor
+  ##   # The code `x[1..2, _][condition] = 1000` would fail with
+  ##   # a `a slice of an immutable tensor cannot be assigned to` error
+  ##   # Instead, using `at_mut` allows assignment to the slice
+  ##   x.at_mut(1..2, _)[condition] = 1000
+  ##
+  ## Input:
+  ##   - a Tensor
+  ##   - and:
+  ##     - specific coordinates (``varargs[int]``)
+  ##     - or a slice (cf. tutorial)
+  ## Returns:
+  ##   - a mutable value or view of the Tensor corresponding to the slice
+  ##     Singleton dimension are collapsed
+  var mt = t[args].squeeze
+  mt
+
+template mut*[T](t: Tensor[T]): var Tensor[T] =
```

Comment:
This is breaking assumptions and should be named unsafe_mut, but I think `at_mut` would cover all use-cases no?

Reply:
I think you are right on both accounts (i.e. it is unsafe and we might not need it).
I was looking for the best way to indicate that the slice can be mutated. `at_mut` probably covers the cases I've uncovered so far so I guess I could remove `mut` for now and we could reconsider it in case we find a case that is not covered by `at_mut`.

That being said, I find that the syntax using `mut` looks better as it does not force you to switch from the `[]=` operator to the `at_mut` function call syntax. If only there was a way to use `mut` but have it fail to compile when the target Tensor were immutable... That is, if we could support:
```nim
var x = arange(12).reshape([4, 3])
x[1..2, _].mut[condition] = 100
```
while this failed to compile:
```nim
let x = arange(12).reshape([4, 3])
x[1..2, _].mut[condition] = 100
```
Then `mut` (which would no longer be unsafe) would be the superior solution IMHO.
If you have any ideas for that let me know, otherwise we can just keep `at_mut`.

Reply:
There are 2 solutions to this.

The first one and the most ergonomic one is to ensure that when slicing `var Tensor` you return a `var Tensor`.

I started and abandoned `var slice` 5 years ago: https://github.com/mratsim/Arraymancer/commit/9b5539079dbf34535a5372f80dad980a0059d990
and iirc try to resurrect them with Nim ``--experimental:views`` support but I was still kind of stuck by the type system, see git blame: https://github.com/mratsim/Arraymancer/blame/f809e7f2806020f4c49ca45f000e478fe53fd2ba/src/arraymancer/tensor/private/p_accessors_macros_write.nim#L262

The second one would be quite a burden. It is actually used internally to distinguish between mutable and immutable memory when using pointers:

You use distinct types and you implement only mutable procs on the "Mut" variant:
https://github.com/mratsim/Arraymancer/blob/f809e7f2806020f4c49ca45f000e478fe53fd2ba/src/arraymancer/laser/tensor/datatypes.nim#L24-L25

This works well for internal pointers because pointer only support indexing reads and indexing writes. For tensors, there are too many functions and I don't thing there is a way to make that "TensorViewImmutable" disappear and becoming a simple "Tensor" when assigned to `let`, i.e. it would only exist in flight.

It should be possible via a macro, say we call it `chain` or `chainMut` and internally we assign to `var` but code becomes quite unintuitive.

Reply:
See also #487